ml-prims/src/common/device_buffer.hpp

```diff
@@ -64,30 +64,30 @@ class device_buffer : public buffer_base<T>
     {
         if ( _size > 0 )
         {
-            _data = static_cast<value_type*>(_allocator->allocate( _capacity*sizeof(value_type), _stream ));
-            CUDA_CHECK( cudaStreamSynchronize( _stream ) );
+            _data = static_cast<value_type*>(_allocator->allocate( _capacity*sizeof(value_type), get_stream() ));
+            CUDA_CHECK( cudaStreamSynchronize( get_stream() ) );
```

Comment:
@jirikraus Sorry, I realize that this question is not related to this PR! Do we need this synchronize step? I suppose, as long as the caller adheres to the stream semantics, we should be fine without it, no?

Reply:
IIRC we already touched on this in a discussion on the PR which introduced this data structure. From my point of view the expectation of a user of `host_buffer` or `device_buffer` is that it is usable after constructions. Thus the synchronization to avoid surprises. However I agree that given we pass in a stream that surprise of some users is something we can probably live with. Btw. if you want a fully asynchronous construction you can do it with:
```cpp
device_buffer tmp( allocator, stream, 0 )
tmp.resize(n,stream);
```
I am undecided what the better option is. Any other thoughts on this?

Reply:
`host_buffer` can also potentially be directly accessed from CPU code. So, it does make sense to put a sync call there to avoid surprises.

However, after a `device_buffer` is constructed, the only way it can be touched is by using a cuda api or a cuda kernel, both of which adhere to stream semantics. Thus, as long as the underlying allocator adheres to the stream semantics, we could live without a sync here. What say?

Reply:
Yes that is a good argument. Only concern I have with that it would be confusing if `host_buffer` and `device_buffer` have different semantics.

Reply:
Agree that this can be confusing to some folks. May be we could document this difference somewhere so that our devs are aware?

Reply:
Sure the behavior should be documented irrespective of what it is. However that does not avoid that either `host_buffer` and `device_buffer` have inconsistent, i.e. confusing, behavior if we remove the stream sync only for `host_buffer` or lead to potential surprises with `host_buffer` if we remove it for both. Anyway I think a way out of this is not providing a constructor that takes a initial size, i.e. it would be used like this:

```cpp
device_buffer tmp( allocator, stream);
tmp.resize(n,stream);
```

With that I think it would be more obvious to users of `host_buffer` or `device_buffer` that the allocation might be asynchronous. What do you think?

Reply:
I really love the convenience of constructing the buffer inside the constructor! One such proposed changes can be found [here](https://github.com/rapidsai/cuml/pull/331/files#diff-a4750fdce2b183579e6c7c36a662b42cR82). As you can see, allocating multiple such device_buffers will become more verbose, if we stop providing initial size in the constructor.

My personal opinion, this inconsistency between `host_buffer` and `device_buffer` should be ok for the sake of performance (especially for the `device_buffer`) as well as the convenience it brings.

Reply:
I think it would be good to gather more opinions on this. @teju85 who else could chime in? Can you tag a few people?

Reply:
Makes sense. I'd start with @jrhemstad and @harrism . I have also requested other ML devs to contribute their thoughts here as well.

Reply:
With the removal of `cudaStreamSynchronize()`, while `device_buffer`s c'tor becomes inconsistent with 'host_buffer's, it is consistent with CUDA APIs/libraries'. I think it's better from the perspective of CUDA echosystem. It would not be problem if we document well in the right place and users have knowledge of stream semantics.

Reply:
I would not worry about making, CPU and GPU allocator consistent, given they are different devices serving different purposes. Also i agree with @minseokl , the removal of sync , makes it consistent with other GPU API and libraries.

Reply:
One issue is that even if the constructor / allocator takes a stream, users might access that memory from a kernel or copy on another stream. It's nice to provide them the convenience of doing so, since they are used to that with cudaMalloc. In fact, even though RMM takes a stream in `rmmAlloc()`, the returned memory is immediately available for use on any stream! The stream is basically an optimization, not a requirement (we usually pass the default stream in cuDF). We get away with this because `rmmAlloc()` is fast and only synchronizes if a) the only available free block is in a heap associated with a different stream than specified (it synchronizes the owner stream and then steals ownership to the specified stream) or b) there are no free blocks and therefore it has to call `cudaMalloc` (which synchronizes everything), and is rare if the initial pool size is sufficient.

I like the guarantee that the memory is immediately valid and available to all streams. I think it probably leads to fewer bugs. And RMM/cnmem makes a good compromise in that MOST allocation is fast and async wrt device work in other streams.

Reply:
To answer with more questions: suppose we skip the synchronize, which has a greater probability to introduce bugs. How difficult is the issue to debug? Will users only see it if they run cuda-memcheck?

Reply:
@NickLeaf the [reserve](https://github.com/rapidsai/cuml/blob/branch-0.7/ml-prims/src/common/device_buffer.hpp#L80) method of `device_buffer` already skips this synchronize call (as expected).  I don't think this `reserve` method, however, has any bugs due to this.

Also, IIUC @harrism's comment above, it seems that it is OK to assume that the underlying custom-device-allocator can also be assumed to follow stream-semantics?

Reply:
BTW, to those new, during the initial PR #247 we did try to discuss about this sync behavior [here](https://github.com/rapidsai/cuml/pull/247#discussion_r260805596).

Reply:
@teju85 that's not what I meant. As I wrote: "I like the guarantee that the memory is immediately valid and available to all streams. I think it probably leads to fewer bugs."

Reply:
@harrism I was referring to your following statement about RMM: `In fact, even though RMM takes a stream in rmmAlloc(), the returned memory is immediately available for use on any stream!`

Reply:
You could have a parameter that you set with the allocator. Call cudaStreamSynchronize() (or cudaDeviceSynchronize()) if the allocator is asynchronous, and don't call it if not.

Reply:
Thanks for all the inputs. I don't think we have come to a conclusion yet. @teju85 what do you think if we move the discussion if the constructor of `device_buffer` should synchronize or not to an issue and move on with this pull request (which as you state initially is unrelated to the open discussion) as it is.

Reply:
Agreed. Doesn't make sense to block this PR for this discussion. Have filed issue #425 and let's migrate this discussion over there.